codegen/src/api/mod.rs

```diff
@@ -0,0 +1,360 @@
+                parse_quote!(::subxt::sp_arithmetic::per_things::Perquintill),
+            ),
+            (
+                "frame_support::traits::misc::WrapperKeepOpaque",
+                parse_quote!(::subxt::WrapperKeepOpaque),
+            ),
+        ]
+        .iter()
+        .map(|(path, substitute): &(&str, syn::TypePath)| {
+            (path.to_string(), substitute.clone())
+        })
+        .collect::<HashMap<_, _>>();
+
+        for (path, substitute) in item_mod_ir.type_substitutes().iter() {
+            type_substitutes.insert(path.to_string(), substitute.clone());
+        }
+
+        let type_gen = TypeGenerator::new(
+            &self.metadata.types,
+            "runtime_types",
+            type_substitutes,
+            derives.clone(),
+        );
+        let types_mod = type_gen.generate_types_mod();
+        let types_mod_ident = types_mod.ident();
+        let pallets_with_mod_names = self
+            .metadata
+            .pallets
+            .iter()
+            .map(|pallet| {
+                (
+                    pallet,
+                    format_ident!("{}", pallet.name.to_string().to_snake_case()),
+                )
+            })
+            .collect::<Vec<_>>();
+        let modules = pallets_with_mod_names.iter().map(|(pallet, mod_name)| {
+            let calls = if let Some(ref calls) = pallet.calls {
+                calls::generate_calls(&type_gen, pallet, calls, types_mod_ident)
+            } else {
+                quote!()
+            };
+
+            let event = if let Some(ref event) = pallet.event {
+                events::generate_events(&type_gen, pallet, event, types_mod_ident)
+            } else {
+                quote!()
+            };
+
+            let storage_mod = if let Some(ref storage) = pallet.storage {
+                storage::generate_storage(&type_gen, pallet, storage, types_mod_ident)
+            } else {
+                quote!()
+            };
+
+            quote! {
+                pub mod #mod_name {
+                    use super::#types_mod_ident;
+                    #calls
+                    #event
+                    #storage_mod
+                }
+            }
+        });
+
+        let outer_event_variants = self.metadata.pallets.iter().filter_map(|p| {
+            let variant_name = format_ident!("{}", p.name);
+            let mod_name = format_ident!("{}", p.name.to_string().to_snake_case());
+            let index = proc_macro2::Literal::u8_unsuffixed(p.index);
+
+            p.event.as_ref().map(|_| {
+                quote! {
+                    #[codec(index = #index)]
+                    #variant_name(#mod_name::Event),
+                }
+            })
+        });
+
+        let outer_event = quote! {
+            #derives
+            pub enum Event {
+                #( #outer_event_variants )*
+            }
+        };
+
+        // todo: [AJ] keep all other code items from decorated mod?
```

Comment:
Should we scrape all these todos and make them into issues/put them into a project? might be a good starting point

Reply:
https://github.com/paritytech/subxt/issues/309